x/mongo/driver/session/client_session.go

```diff
@@ -215,16 +215,31 @@ func NewClientSession(pool *Pool, clientID uuid.UUID, sessionType Type, opts ...
 		return nil, errors.New("causal consistency and snapshot cannot both be set for a session")
 	}
 
-	servSess, err := pool.GetSession()
-	if err != nil {
-		return nil, err
+	if sessionType != Implicit || !mergedOpts.deferImplicit {
+		if err := c.SetServer(); err != nil {
+			return nil, err
+		}
 	}
 
-	c.Server = servSess
-
 	return c, nil
 }
 
+// NewClientImplicitSessionDeferred will create a new implicit session client, defering the server checkout as a
+// user-defined operation.
+func NewClientImplicitSessionDeferred(pool *Pool, clientID uuid.UUID, opts ...*ClientOptions) (*Client, error) {
```

Comment:
If we get rid of `deferImplicit`, I don't think this new constructor is necessary.

Reply:
See [this comment](https://github.com/mongodb/mongo-go-driver/pull/1027#discussion_r929121724). I went back and forth with this a lot, IMO it's better practice to preserve the existing functionality of `NewClientSession`. It's used in quite a few places and still used in some of the collection CRUD methods, such as `drop`.